src/devices/Gpio/Drivers/Rockchip/RockchipDriver.cs

```diff
@@ -0,0 +1,282 @@
+                        SetPinMode(pinNumber, PinMode.Input);
+                        break;
+                    case PinMode.Output:
+                        Write(pinNumber, PinValue.Low);
+                        SetPinMode(pinNumber, PinMode.Input);
+                        break;
+                    default:
+                        break;
+                }
+
+                _pinModes.Remove(pinNumber);
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void SetPinMode(int pinNumber, PinMode mode)
+        {
+            // different chips have different number of GRFs and offsets
+            // this method needs to be overridden in subclasses
+            base.OpenPin(pinNumber);
+            base.SetPinMode(pinNumber, mode);
+
+            if (_pinModes.ContainsKey(pinNumber))
+            {
+                _pinModes[pinNumber].CurrentPinMode = mode;
+            }
+            else
+            {
+                _pinModes.Add(pinNumber, new PinState(mode));
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void Write(int pinNumber, PinValue value)
+        {
+            (int GpioNumber, int Port, int PortNumber) unmapped = UnmapPinNumber(pinNumber);
+
+            // data register (GPIO_SWPORT_DR) offset is 0x0000
+            uint* dataPointer = (uint*)_gpioPointers[unmapped.GpioNumber];
+            uint dataValue = *dataPointer;
+
+            if (value == PinValue.High)
+            {
+                dataValue |= 0b1U << (unmapped.Port * 8 + unmapped.PortNumber);
+            }
+            else
+            {
+                dataValue &= ~(0b1U << (unmapped.Port * 8 + unmapped.PortNumber));
+            }
+
+            *dataPointer = dataValue;
+        }
+
+        /// <inheritdoc/>
+        protected unsafe override PinValue Read(int pinNumber)
+        {
+            (int GpioNumber, int Port, int PortNumber) unmapped = UnmapPinNumber(pinNumber);
+
+            // data register (GPIO_EXT_PORTA) offset is 0x0050
+            uint* dataPointer = (uint*)(_gpioPointers[unmapped.GpioNumber] + 0x0050);
+            uint dataValue = *dataPointer;
+
+            return Convert.ToBoolean((dataValue >> (unmapped.Port * 8 + unmapped.PortNumber)) & 0b1) ? PinValue.High : PinValue.Low;
+        }
+
+        /// <inheritdoc/>
+        protected override void AddCallbackForPinValueChangedEvent(int pinNumber, PinEventTypes eventTypes, PinChangeEventHandler callback)
+        {
+            _pinModes[pinNumber].InUseByInterruptDriver = true;
+
+            base.OpenPin(pinNumber);
+            base.AddCallbackForPinValueChangedEvent(pinNumber, eventTypes, callback);
+        }
+
+        /// <inheritdoc/>
+        protected override void RemoveCallbackForPinValueChangedEvent(int pinNumber, PinChangeEventHandler callback)
+        {
+            _pinModes[pinNumber].InUseByInterruptDriver = false;
+
+            base.OpenPin(pinNumber);
```

Comment:
Why would you need to open the pin again? you are already doing that when adding the callback right?